lite2/client.go

```diff
@@ -0,0 +1,374 @@
+	}
+
+	// 3) Persist both of them and continue.
+	return c.updateTrustedHeaderAndVals(h, vals)
+}
+
+// SetLogger sets a logger.
+func (c *Client) SetLogger(l log.Logger) {
+	c.logger = l
+}
+
+// TrustedHeader returns a trusted header at the given height or nil if no such
+// header exist. It returns an error if there are some issues with the trusted
+// store, although that should not happen normally. TODO mention how many
+// headers will be kept by the light client.
+//
+// 0 - the latest.
+// height must be >= 0.
+func (c *Client) TrustedHeader(height int64) (*types.SignedHeader, error) {
+	if height < 0 {
+		return nil, errors.New("negative height")
+	}
+
+	if height == 0 {
+		var err error
+		height, err = c.LastTrustedHeight()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return c.trustedStore.SignedHeader(height)
+}
+
+// LastTrustedHeight returns a last trusted height.
+func (c *Client) LastTrustedHeight() (int64, error) {
+	return c.trustedStore.LastSignedHeaderHeight()
+}
+
+// ChainID returns a chain ID.
+func (c *Client) ChainID() string {
+	return c.chainID
+}
+
+// VerifyHeaderAtHeight fetches the header and validators at the given height
+// and calls VerifyHeader.
+//
+// If the trusted header is more recent than one here, an error is returned.
+func (c *Client) VerifyHeaderAtHeight(height int64, now time.Time) error {
+	if c.trustedHeader.Height >= height {
+		return errors.Errorf("height #%d is already trusted (last: #%d)", height, c.trustedHeader.Height)
+	}
+
+	// Request the header and the vals.
+	newHeader, newVals, err := c.fetchHeaderAndValsAtHeight(height)
+	if err != nil {
+		return err
+	}
+
+	return c.VerifyHeader(newHeader, newVals, now)
+}
+
+// VerifyHeader verifies new header against the trusted state.
+//
+// SequentialVerification: verifies that 2/3 of the trusted validator set has
+// signed the new header. If the headers are not adjacent, **all** intermediate
+// headers will be requested.
+//
+// SkippingVerification(trustLevel): verifies that {trustLevel} of the trusted
+// validator set has signed the new header. If it's not the case and the
+// headers are not adjacent, bisection is performed and necessary (not all)
+// intermediate headers will be requested. See the specification for the
+// algorithm.
+//
+// If the trusted header is more recent than one here, an error is returned.
+func (c *Client) VerifyHeader(newHeader *types.SignedHeader, newVals *types.ValidatorSet, now time.Time) error {
+	if c.trustedHeader.Height >= newHeader.Height {
+		return errors.Errorf("height #%d is already trusted (last: #%d)", newHeader.Height, c.trustedHeader.Height)
+	}
+
+	var err error
+	switch c.mode {
+	case sequential:
+		err = c.sequence(newHeader, newVals, now)
+	case skipping:
+		err = c.bisection(c.trustedHeader, c.trustedNextVals, newHeader, newVals, now)
+	}
+	if err != nil {
+		return err
+	}
+
+	// Update trusted header and vals.
+	nextVals, err := c.primary.ValidatorSet(newHeader.Height + 1)
```

Comment:
Does this function verify against the trusted validator set hash? Below it looks like we call it and then verify.

I think we do need to verify it here, right? Does `updateTrustedHeaderAndVals` do that?

I think functions which verify should include "verify" in their names.

Reply:
good catch!